enterprise/ssl-impl/src/main/java/io/crate/protocols/ssl/FilesWatcher.java

```diff
@@ -0,0 +1,130 @@
+import java.util.function.Consumer;
+
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+/**
+ * The file watcher checks for changes on registered files
+ * and notifies their listeners accordingly.
+ */
+class FilesWatcher implements Runnable, Closeable {
+
+    private final WatchService watchService;
+    private final Map<WatchKey, List<Watcher>> watchers = new HashMap<>();
+
+    FilesWatcher() throws IOException {
+        this.watchService = FileSystems.getDefault().newWatchService();
+    }
+
+    void addListener(Path path,
+                     Consumer<WatchEvent<Path>> listener,
+                     WatchEvent.Kind<?>... watchEventKinds) throws IOException {
+        Watcher watcher = new Watcher(path, listener);
+        WatchKey key = watcher.registrablePath().register(watchService, watchEventKinds);
+        if (watchers.containsKey(key)) {
+            watchers.get(key).add(watcher);
+        } else {
+            watchers.put(key, new ArrayList<>(Collections.singletonList(watcher)));
+        }
+    }
+
+    @Override
+    public void run() {
+        poll();
+    }
+
+    /**
+     * Polls for watch events on the registered paths.
+     */
+    private void poll() {
+        // break the loop if no watch keys are present in the queue
+        // or watch key events were consumed
+        while (true) {
+            WatchKey key = watchService.poll();
+            if (key == null || !watchers.containsKey(key)) {
+                break;
+            }
+
+            for (WatchEvent<?> event : key.pollEvents()) {
+                WatchEvent.Kind<?> kind = event.kind();
+                if (OVERFLOW == kind) {
+                    continue;
+                }
+                //noinspection unchecked
+                WatchEvent<Path> watchEvent = (WatchEvent<Path>) event;
+                Path modifiedPath = watchEvent.context();
+                for (var watcher : watchers.get(key)) {
+                    if (watcher.path.toFile().getName().equals(modifiedPath.toFile().getName())) {
+                        watcher.listener.accept(watchEvent);
+                    }
+                }
+
+                boolean valid = key.reset();
+                if (!valid) {
+                    break;
+                }
+            }
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        watchService.close();
+        watchers.clear();
+    }
+
+    private static class Watcher {
+        final Consumer<WatchEvent<Path>> listener;
+        final Path path;
+
+        private Watcher(Path path, Consumer<WatchEvent<Path>> listener) {
+            this.path = path;
+            this.listener = listener;
+        }
+
+        /**
+         * Returns a path that can be registered with the watch service.
+         * <p>
+         * Only directories can be registered with the watch service.
+         */
+        Path registrablePath() {
+            if (Files.isDirectory(path)) {
```

Comment:
I wonder if we should throw an error in case the path isn't a directory instead of implictly falling back to the parent.

E.g. a user of the API might assume that the path is to a file directly. In which case the user could be inclined to register multiple files that are all within the same directory. So the same watchService will end up being registered multiple times under the same path. (Which is what we're doing)

Another option would be to turn around `https://github.com/crate/crate/pull/8755/files#diff-9b78cba2eda0566273f92489a63182e2R45` and avoid a register call but just extend the list of (sub) paths that are to be monitored within a path.

Reply:
Actually, that was kind of an intention for the API to allow users to provide full paths to files that are to be monitored regardless whether they reside within the same directory or not. That's why we do the filtering on file names (obtained from WatchEvent context) before calling files listeners in the poll method and keep the monitored files paths under the same WatchKey, so we can manipulate with the map of registered paths easily when we receive a watch key in the poll method. We can probably avoid a call for registering the same path with the watch service, but that would just shift complexity to another place where we would need to do some work on checking relations between paths, etc.

I am not sure about the part with raising an exception. Do you have some other idea of API where we can register only paths to directories with cert stores?

Reply:
Ftr: The issue I described isn't an issue because the register/watchService take care of this. If the same directory is registered multiple times one always gets back the same watchKey instance.